python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1380 @@
+        """
+        sigma = self.config.nSigmaClipGainCalc
+        maxLag = self.config.maxLag
+        border = self.config.nPixBorderGainCalc
+        biasCorr = self.config.biasCorr
+
+        # NB: don't use dataRef.get('raw_detector') due to composites
+        detector = dataRef.get('camera')[dataRef.dataId[self.config.ccdKey]]
+
+        ampMeans = {}
+
+        dataRef.dataId['visit'] = v1
+        exp1 = self.isr.runDataRef(dataRef).exposure
+        dataRef.dataId['visit'] = v2
+        exp2 = self.isr.runDataRef(dataRef).exposure
+        del dataRef.dataId['visit']
+        exps = [exp1, exp2]
+        self._checkExpLengthEqual(exp1, exp2, v1, v2)
+
+        detector = exps[0].getDetector()  # note we lose the detector in the next line as they belong to exp
+        ims = [self._convertImagelikeToFloatImage(exp) for exp in exps]
+
+        if self.debug.display:
+            self.disp1.mtv(ims[0], title=str(v1))
+            self.disp2.mtv(ims[1], title=str(v2))
+
+        sctrl = afwMath.StatisticsControl()
+        sctrl.setNumSigmaClip(sigma)
+        for imNum, im in enumerate(ims):
+
+            # calculate the sigma-clipped mean, excluding the borders
+            # safest to apply borders to all amps regardless of edges
+            # easier, camera-agnostic, and mitigates potentially dodgy
+            # overscan-biases around edges as well
+            for amp in detector:
+                ampName = amp.getName()
+                ampIm = im[amp.getBBox()]
+                mean = afwMath.makeStatistics(ampIm[border:-border, border:-border, afwImage.LOCAL],
+                                              afwMath.MEANCLIP, sctrl).getValue()
+                if ampName not in ampMeans.keys():
+                    ampMeans[ampName] = []
+                ampMeans[ampName].append(mean)
+                ampIm -= mean
+
+        diff = ims[0].clone()
+        diff -= ims[1]
+
+        temp = diff[border:-border, border:-border, afwImage.LOCAL]
+
+        # Subtract background. It should be a constant,
+        # but it isn't always (e.g. some SuprimeCam flats)
+        # TODO: Check how this looks, and if this is the "right" way to do this
+        binsize = self.config.backgroundBinSize
+        nx = temp.getWidth()//binsize
+        ny = temp.getHeight()//binsize
+        bctrl = afwMath.BackgroundControl(nx, ny, sctrl, afwMath.MEANCLIP)
+        bkgd = afwMath.makeBackground(temp, bctrl)
+
+        box = diff.getBBox()
+        box.grow(-border)
+        diff[box, afwImage.LOCAL] -= bkgd.getImageF(afwMath.Interpolate.CUBIC_SPLINE,
+                                                    afwMath.REDUCE_INTERP_ORDER)
+
+        variances = {}
+        coVars = {}
+        for amp in detector:
+            ampName = amp.getName()
+
+            diffAmpIm = diff[amp.getBBox()].clone()
+            diffAmpImCrop = diffAmpIm[border:-border-maxLag, border:-border-maxLag, afwImage.LOCAL]
```

Comment:
Please add spaces after ":" and around "-" (or before, in the case of negating `border`). I suggest a global regex search for + and - without surrounding spaces

Reply:
Done! (I think)